Lib/asyncio/windows_events.py

```diff
@@ -324,13 +324,13 @@ def _run_forever_cleanup(self):
         if self._self_reading_future is not None:
             ov = self._self_reading_future._ov
             self._self_reading_future.cancel()
-            # self_reading_future was just cancelled so if it hasn't been
-            # finished yet, it never will be (it's possible that it has
-            # already finished and its callback is waiting in the queue,
-            # where it could still happen if the event loop is restarted).
-            # Unregister it otherwise IocpProactor.close will wait for it
-            # forever
-            if ov is not None:
+            # self_reading_future always uses IOCP, so even though it's
+            # been cancelled, we need to make sure that the IOCP message
+            # is received so that the kernel is not holding on to the
+            # memory, possibly causing memory corruption later. Only
+            # unregister it if IO is complete in all respects. Otherwise
+            # we need another _poll() later to complete the IO.
+            if ov is not None and not ov.pending:
```

Comment:
Yes, that's the case. Issuing the cancel starts the process of cancelling the IO in the Win32 API. Since this is a request and asynchronous, it's not immediate. The outcome of this can either be that the `pending` flag changes to `False`, or an overlapped IO message was already queued and `pending` will never change to `False` until the message is read.

The previous code was not checking the `pending` flag and always allowing the `_overlapped` object to be destroyed even if it was still in use. By now checking the `pending` flag we allow it to be retained until the IO is read "later" (typically the next call to `run_forever`() or upon calling `close()` on the event loop).

New unit test was added to ensure that the event loop exits correctly (otherwise test would hang) and ensure that the error message in question is never thrown.

Reply:
@zooba @gvanrossum what else do we need to get this wrapped up?